### ISSUE TITLE
Potential weakness

### ISSUE THREAD

**Issue by tevador** [hyc/randprog#5]
Currently, the algorithm generates a javascript program that looks like a C program and uses only a handful of basic operations. This means that an efficient runtime can be created in software (or hardware) which only implements a subset of the javscript VM. This could be used to gain an advantage in mining performance.

I suggest to modify the generation routine to include some of the following in the program:
- built-in JS functions (such as String.length or Array.forEach)
- objects
- closures
- anonymous functions
- function variables
- function prototypes
- typeof, instanceof operators

This would make it more likely that miners have to use a full javascript VM.

**Comment by hyc** [comment]
Agreed. I've been considering which of the built in libraries to add to the Function selector. Math.xx was my first candidate but there's some concern that accuracy will differ across platforms or implementations. Still, the Javascript specification states that all numbers are "double-precision 64-bit binary format IEEE 754-2008 value" so they should actually be uniform across conforming Javascript implementations.

Anyway, as far as adding features - patches welcome.

**Comment by tevador** [comment]
From the IEEE 754-2008 specification, chapter 9 says that the precision of named functions (such as sqrt or log) is language-defined.

The Javascript specification says it's [implementation-defined](https://www.ecma-international.org/ecma-262/6.0/#sec-function-properties-of-the-math-object).

**Comment by hyc** [comment]
OK. I guess we can only touch stuff like abs, floor, etc that have well defined results.
